problemtools/verifyproblem.py

```diff
@@ -740,6 +741,296 @@ def check(self, args):
+        self.configfile = os.path.join(problem.probdir, 'generators', 'generators.yaml')
+        self._data = None
+        self._testcases = []
+        self._testdata_yaml = {}
+        self._data_directories = set()
+
+        if os.path.isfile(self.configfile):
+            try:
+                with open(self.configfile) as f:
+                    self._data = yaml.safe_load(f)
+                # Loading empty yaml yields None, for no apparent reason...
+                if self._data is None:
+                    self._data = {}
+            except Exception as e:
+                self.error(e)
+
+        if isinstance(self._data, dict):
+            # The top-level dict always represents a directory, even if there
+            # is no type key
+            self._data['type'] = 'directory'
+
+    def __str__(self):
+        return 'generators'
+
+    def _parse_command(self, key, state):
+        command = state[key]
+        name = os.path.basename(state['path'])
+        random_salt = str(state['random_salt'])
+
+        def err():
+            self.error('Invalid %s key for path %s in generators.yaml' % (key, state['path']))
+
+        if not isinstance(command, str):
+            return err()
+
+        seed = str(int(hashlib.sha512((random_salt + command).encode('utf-8')).hexdigest(), 16) % (2**31))
+
+        parts = shlex.split(command)
+        if not parts:
+            return err()
+
+        for i, part in enumerate(parts):
+            new = ''
+            for j, group in enumerate(part.split('{')):
+                if group.count('}') != (0 if j == 0 else 1):
+                    return err()
+                if j == 0:
+                    new += group
+                else:
+                    group, rest = group.split('}')
+                    if group.startswith('seed'):
+                        new += seed
+                    elif group == 'name':
+                        new += name
+                    else:
+                        return err()
+                    new += rest
+            parts[i] = new
+
+        program, arguments = parts[0], parts[1:]
+        if program not in self._generators:
+            self._generators[program] = program
+
+        return (program, arguments)
+
+    def _parse_testcase(self, data, state):
+        if state['input'] is None:
+            self.error('Path %s in generators.yaml must contain an input key' % state['path'])
+        for key in ['input', 'solution', 'visualizer']:
+            if state[key] is not None:
+                state[key] = self._parse_command(key, state)
+        if state['input'] is not None:
+            self._testcases.append(state)
+
+    def _parse_directory(self, data, state):
+        # TODO: Process includes
+
+        if 'testdata.yaml' in data:
+            content = data['testdata.yaml']
+            if content is None:
+                content = {}
+            self._testdata_yaml['%s/%s' % (state['path'], 'testdata.yaml')] = content
```

Comment:
These kind of things are much nicer using `pathlib.Path` IMO.

Reply:
I identify particular entries within the generators.yaml file with a slash-separated sequence of keys, e.g. `data: sample: solution: test.cpp` is identified with `data/sample/solution`. This is then used in error messages and other places. While this may in many cases correspond to a particular file, it won't always, so it felt wrong to be using `pathlib` here.